pkg/issuer/vault/issue.go

```diff
@@ -0,0 +1,286 @@
+	}
+
+	commonName := crt.Spec.CommonName
+	altNames := crt.Spec.DNSNames
+	if len(commonName) == 0 && len(altNames) == 0 {
+		return nil, nil, fmt.Errorf("no domains specified on certificate")
+	}
+
+	crtPem, err := v.signCertificate(crt, signeeKey)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return pki.EncodePKCS1PrivateKey(signeeKey), crtPem, nil
+}
+
+// signCertificate returns a signed x509.Certificate object for the given
+// *v1alpha1.Certificate crt.
+func (v *Vault) signCertificate(crt *v1alpha1.Certificate, key *rsa.PrivateKey) ([]byte, error) {
+	commonName := pki.CommonNameForCertificate(crt)
+	altNames := pki.DNSNamesForCertificate(crt)
+
+	if len(commonName) == 0 && len(altNames) > 0 {
+		commonName = altNames[0]
+	}
+
+	template := pki.GenerateCSR(commonName, altNames...)
+	template.Subject.Organization = []string{defaultOrganization}
+
+	derBytes, err := x509.CreateCertificateRequest(rand.Reader, template, key)
+	if err != nil {
+		return nil, fmt.Errorf("error creating x509 certificate: %s", err.Error())
+	}
+
+	pemRequestBuf := &bytes.Buffer{}
+	err = pem.Encode(pemRequestBuf, &pem.Block{Type: "CERTIFICATE REQUEST", Bytes: derBytes})
+	if err != nil {
+		return nil, fmt.Errorf("error encoding certificate request: %s", err.Error())
+	}
+
+	return v.requestVaultCert(commonName, altNames, pemRequestBuf.String())
+}
+
+func (v *Vault) initVaultClient() (*vault.Client, error) {
+	client, err := vault.NewClient(nil)
+	if err != nil {
+		return nil, fmt.Errorf("error initializing Vault client: %s", err.Error())
+	}
+
+	client.SetAddress(v.issuer.GetSpec().Vault.Server)
+
+	tokenRef := v.issuer.GetSpec().Vault.Auth.TokenSecretRef
+	if tokenRef.Name != "" {
+		token, err := v.vaultTokenRef(tokenRef.Name, tokenRef.Key)
+		if err != nil {
+			return nil, fmt.Errorf("error reading Vault token from secret %s/%s: %s", v.issuerResourcesNamespace, tokenRef.Name, err.Error())
+		}
+		client.SetToken(token)
+
+		return client, nil
+	}
+
+	appRole := v.issuer.GetSpec().Vault.Auth.AppRole
+	if appRole.RoleId != "" {
+		token, err := v.requestTokenWithAppRoleRef(client, &appRole)
+		if err != nil {
+			return nil, fmt.Errorf("error reading Vault token from secret %s/%s: %s", v.issuerResourcesNamespace, appRole.SecretRef.Name, err.Error())
+		}
+		client.SetToken(token)
+
+		return client, nil
+	}
+
+	return nil, fmt.Errorf("error initializing Vault client. tokenSecretRef or appRoleSecretRef not set")
+}
+
+func (v *Vault) requestTokenWithAppRoleRef(client *vault.Client, appRole *v1alpha1.VaultAppRole) (string, error) {
+	roleId, secretId, err := v.appRoleRef(appRole)
+	if err != nil {
+		return "", fmt.Errorf("error reading Vault AppRole from secret: %s/%s: %s", appRole.SecretRef.Name, v.issuerResourcesNamespace, err.Error())
+	}
+
+	parameters := map[string]string{
+		"role_id":   roleId,
+		"secret_id": secretId,
+	}
+
+	url := "/v1/auth/approle/login"
+
+	request := client.NewRequest("POST", url)
```

Comment:
Approle auth has to be done with a raw client? 😢

Reply:
I don't find any approle helpers in the Vault api. I don't want to use inner parts of Vault to not be liable to big changes on their side too...